src/storage/cache_proxy/cluster_manager.rs

```diff
@@ -0,0 +1,1096 @@
+            format!(
+                "Current node {:?} Failed to watch the node list update, and update the ring",
+                current_node_info.ip()
+            )
+        })?;
+
+        Ok(())
+    }
+
+    /// Do slave tasks
+    ///
+    /// Slave node will watch the ring update and campaign master
+    #[allow(clippy::arithmetic_side_effects, clippy::pattern_type_mismatch)] // The `select!` macro will generate code that goes against these rules.
+    async fn do_slave_tasks(&self) -> DatenLordResult<()> {
+        // Try to watch master and hashring
+        // Wait for status update
+        let current_node_info = self.inner.get_node();
+        debug!(
+            "do_slave_tasks: Node {:?} will watch the ring update and campaign master",
+            current_node_info.ip()
+        );
+        // Check the node status
+        if current_node_info.status() != NodeStatus::Slave {
+            return Ok(());
+        }
+
+        // Block here, try to watch master
+        // If the master is down and return ok, the slave node will try to get the master lock
+        // If watch_master failed, we will retry to watch master
+        self.inner.watch_master().await?;
+
+        Ok(())
+    }
+}
+
+/// Inner struct of cluster manager
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct ClusterManagerInner {
+    /// Etcd client
+    kv_engine: Arc<KVEngineType>,
+    /// Node session, try to keep the session alive, we will set the session in run method
+    node_session: ArcSwapOption<Session>,
+    /// current node info
+    node: ArcSwap<Node>,
+    /// current node list, we will set the session in run method
+    node_list: ArcSwap<Vec<Node>>,
+    /// current hash ring, we will set the session in run method
+    hash_ring: ArcSwap<Ring<Node>>,
+    // TODO: Support distribute cluster config here
+}
+
+#[allow(dead_code)]
+impl ClusterManagerInner {
+    /// Create a new cluster manager
+    pub fn new(kv_engine: Arc<KVEngineType>, node: Node) -> Self {
+        let node_session = ArcSwapOption::empty();
+        let node = ArcSwap::<Node>::new(Arc::new(node));
+        let node_list = ArcSwap::<Vec<Node>>::new(Arc::new(Vec::new()));
+        let hash_ring = ArcSwap::<Ring<Node>>::new(Arc::new(Ring::default()));
+
+        Self {
+            kv_engine,
+            node_session,
+            node,
+            node_list,
+            hash_ring,
+        }
+    }
+
+    /// Prepare for the session and other data
+    pub async fn prepare(&self) -> DatenLordResult<()> {
+        // Prepare session
+        let session = self
+            .kv_engine
+            .create_session(SESSION_TIMEOUT_SEC)
+            .await
+            .with_context(|| format!("{} failed to create session", self.node.load().ip()))?;
+        self.node_session.store(Some(session));
+        Ok(())
+    }
+
+    /// Update node info
+    pub async fn update_node_in_cluster(&self) -> DatenLordResult<()> {
```

Comment:
If not exposed outside this file,  do not use pub

Reply:
I have removed this function.